test/testdata/infer/unreachable_code_class_variable.rb

```diff
@@ -0,0 +1,15 @@
+# typed: true
+
+extend T::Sig
+
+class C
+  @@x = T.let(42, Integer)
+
+  def f
+    @@x ||= 43 # error: This code is unreachable
```

Comment:
I haven't looked at the other tests that exist for this stuff, but do you think it's worth adding a test like

```ruby
@@y = T.let(0, T.any(FalseClass, Integer)

def self.h
  @@y ||= 1
end
```

Reply:
i.e., something that doesn't conflate "undefined" and "defined but `nil`" but still wants to use `||=`?

(because the unreachable code error would prevent the compiler from activating)

Reply:
Sorbet [doesn't give an error](https://sorbet.run/#%23%20typed%3A%20true%0A%0A%23%20Test%20that%20Sorbet%20correctly%20infers%20the%20types%20of%20class%20variables%20and%20instance%0A%23%20variables%20when%20they%20are%20used%20as%20the%20LHS%20of%20%7C%7C%3D.%0A%0Aclass%20C%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20%40%40never_falsy%20%3D%20T.let%280%2C%20T.any%28FalseClass%2C%20Integer%29%29%0A%0A%20%20sig%20%7Bparams%28x%3A%20Integer%29.returns%28Integer%29%7D%0A%20%20def%20self.f6%28x%29%0A%20%20%20%20%40%40never_falsy%20%7C%7C%3D%20x%20%23%20error%3A%20This%20code%20is%20unreachable%0A%20%20end%0A%0A%20%20sig%20%7Bparams%28x%3A%20Integer%29.returns%28Integer%29%7D%0A%20%20def%20h6%28x%29%0A%20%20%20%20%40%40never_falsy%20%7C%7C%3D%20x%20%23%20error%3A%20This%20code%20is%20unreachable%0A%20%20end%0Aend%0A) on this today.  I don't know if it's worth adding a test.